app/services/moderator/merge_user.rb

```diff
@@ -55,45 +58,77 @@ def update_social
-      @delete_user.mentions.update_all(user_id: @keep_user.id) if @delete_user.mentions.any?
+    def merge_badge_achievements
+      duplicate_badges = @delete_user.badge_achievements.pluck(:badge_id).intersection(
+        @keep_user.badge_achievements.pluck(:badge_id),
+      )
+      @delete_user.badge_achievements.where(badge_id: duplicate_badges).delete_all
+      @delete_user.badge_achievements.update_all(user_id: @keep_user.id)
+      BadgeAchievement.counter_culture_fix_counts(where: { users: { id: @keep_user.id } })
     end
 
-    def merge_follows
-      @delete_user.follows&.update_all(follower_id: @keep_user.id) if @delete_user.follows.any?
-      @delete_user_followers = Follow.followable_user(@delete_user.id)
-      @delete_user_followers.update_all(followable_id: @keep_user.id) if @delete_user_followers.any?
+    def merge_chat_channels
+      duplicate_moderator_memberships = @delete_user.chat_channel_memberships.where(
+        status: "active", role: "mod",
+      ).pluck(:chat_channel_id).intersection(
+        @keep_user.chat_channel_memberships.where(
+          status: "active", role: "mod",
+        ).pluck(:chat_channel_id),
+      )
+      @delete_user.chat_channel_memberships.where(chat_channel_id: duplicate_moderator_memberships).delete_all
+
+      moderator_memberships = @delete_user.chat_channel_memberships.where(
+        status: "active", role: "mod",
+      ).pluck(:chat_channel_id)
+
+      duplicate_memberships = @delete_user.chat_channel_memberships.where(
+        status: "active",
+      ).pluck(:chat_channel_id).intersection(
+        @keep_user.chat_channel_memberships.where(status: "active").pluck(:chat_channel_id),
+      )
+      @delete_user.chat_channel_memberships.where(chat_channel_id: duplicate_memberships).delete_all
+      @keep_user.chat_channel_memberships.where(chat_channel_id: moderator_memberships).update_all(role: "mod")
+      @delete_user.chat_channel_memberships.update_all(user_id: @keep_user.id)
     end
 
-    def merge_content
-      merge_reactions if @delete_user.reactions.any?
-      merge_comments if @delete_user.comments.any?
-      merge_articles if @delete_user.articles.any?
+    def merge_follows
+      duplicate_followers = Follow.where(followable: @delete_user).pluck(:follower_id, :follower_type).intersection(
+        Follow.where(followable: @keep_user).pluck(:follower_id, :follower_type),
+      )
+      Follow.where(follower: duplicate_followers, followable: @delete_user).delete_all
+      Follow.where(followable: @delete_user).update_all(followable_id: @keep_user.id)
+
+      duplicate_followables = Follow.where(follower: @delete_user).pluck(:followable_id, :followable_type).intersection(
+        Follow.where(follower: @keep_user).pluck(:followable_id, :followable_type),
+      )
+      Follow.where(followable: duplicate_followables, follower: @delete_user).delete_all
+      Follow.where(follower: @delete_user).update_all(follower_id: @keep_user.id)
     end
```

Comment:
Similar case here: one additional scope -- `follower` and `followable`. It's otherwise straightforward: handle one case, then handle the other.

Reply:
my clever "just pass the other relation to where" technique falls down a little on the polymorphic relation for follower and followable here (where you group by a pair of id,type). I think the only way out of that hole is to do what you did, otherwise iterate over the follower types, or hard code the tuple sql join (unless you've seen an AR way to do a `WHERE (col1, col2) IN (SELECT col1, col2 FROM ...` style query).

This looks like the same result as duplicate_followers for delete_user.id 15, keep user id 12?

```
SELECT  * FROM follows 
  WHERE followable_id = 15 
  AND followable_type = 'User' 
  AND (follower_id, follower_type) IN (
    SELECT follower_id, follower_type FROM follows 
    WHERE followable_id = 12 
    AND followable_type = 'User'
);
```

Reply:
https://github.com/forem/forem/pull/13275/files/a042571a3af29192c6504c78806c4a6d1647e048#diff-cc0dbafb7a942d6ea1bdd1e295586513bfbceeb150e5e846621bd885dc5695f6R67-R75

 well I guess it's not unheard of - here is rhymes doing a similar subquery select  (probably built smarter than my raw sql statement with that fancy to_sql he added).